lib/nrf9160_time/nrf9160_time.c

```diff
@@ -0,0 +1,284 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <nrf9160_time.h>
+#include <zephyr.h>
+#include <zephyr/types.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <at_cmd.h>
+#include <time.h>
+#include <errno.h>
+#include <string.h>
+#include <net/sntp.h>
+#include <net/socketutils.h>
+#include <sys/timeutil.h>
+
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(nrf9160_time, CONFIG_NRF9160_TIME_LOG_LEVEL);
+
+#define AT_CMD_MODEM_DATE_TIME			"AT+CCLK?"
+#define AT_CMD_MODEM_DATE_TIME_RESPONSE_LEN	28
+
+#define UIO_IP      "129.240.2.6"
+#define GOOGLE_IP   "216.239.35.0"
+#define GOOGLE_IP_2 "216.239.35.4"
+#define GOOGLE_IP_3 "216.239.35.8"
+#define GOOGLE_IP_4 "216.239.35.12"
+
+K_SEM_DEFINE(fetch_time_sem, 0, 1);
+
+static struct k_delayed_work time_work;
+
+static struct time_aux {
+	s64_t date_time_utc;
+	int last_date_time_update;
+} time_aux;
+
+static struct sntp_time sntp_time;
+static bool initial_valid_time;
+
+struct ntp_servers {
+	const char *server;
+};
+
+struct ntp_servers servers[] = {
+	{.server = UIO_IP},
+	{.server = GOOGLE_IP},
+	{.server = GOOGLE_IP_2},
+	{.server = GOOGLE_IP_3},
+	{.server = GOOGLE_IP_4}
+};
+
+#if defined(CONFIG_NRF9160_TIME_MODEM)
+static int parse_time_entries(char *datetime_string, int min, int max)
+{
+	char buf[50];
+
+	for (int i = min; i < max + 1; i++) {
+		buf[i - min] = datetime_string[i];
+	}
+
+	return strtol(buf, NULL, 10);
+}
+
+static int get_time_cellular_network(void)
+{
+	int err;
+	char buf[AT_CMD_MODEM_DATE_TIME_RESPONSE_LEN + 5];
+	struct tm date_time;
+
+	err = at_cmd_write(AT_CMD_MODEM_DATE_TIME, buf, sizeof(buf), NULL);
+	if (err) {
+		LOG_DBG("Could not get cellular network time, error: %d", err);
+		return err;
+	}
+	buf[AT_CMD_MODEM_DATE_TIME_RESPONSE_LEN] = 0;
+
+	LOG_DBG("Response from modem: %s", log_strdup(buf));
+
```

Comment:
An example of a datetime string would be in order here to make it easier to understand.